### PR TITLE
Integration test for the Resource Service and Controller Runtime 🧪

### PR DESCRIPTION
**Pull request by boxofrad**
### Description

Adds an `enable_dev_resource` config flag that can be used to register the demo resources and controllers when not running in `-dev` mode (which isn't really compatible with the integration test framework's config).

Exposes a gRPC client connection on the test framework's `Agent` type.

Introduces an integration test that exercises the full Resource Service and Controller Runtime stack by writing a demo resource and observing the controller's response.

### CODE REVIEW

File: test/integration/consul-container/libs/cluster/container.go [hunk 1]
```diff
@@ -232,6 +239,28 @@ func NewConsulContainer(ctx context.Context, config Config, cluster *Cluster, po
 		info.CACertFile = clientCACertFile
 	}
 
+	// TODO: Support gRPC+TLS port.
```
Comment (boxofrad) [inline test/integration/consul-container/libs/cluster/container.go:242]: I started working on this but couldn't figure out a way to make `UseGRPCWithTLS = true`. Would appreciate some help from the CTIA team here 🙇🏻‍♂️

File: test/integration/consul-container/libs/cluster/network.go [hunk 2]
```diff
@@ -20,6 +20,7 @@ func createNetwork(t TestingT, name string) (testcontainers.Network, error) {
 			Name:           name,
 			Attachable:     true,
 			CheckDuplicate: true,
+			SkipReaper:     isRYUKDisabled(),
```
Comment (boxofrad) [inline test/integration/consul-container/libs/cluster/network.go:23]: This fixes a problem I've been seeing locally where RYUK would still run even when setting the env var to turn it off.

File: agent/config/config.go [hunk 3]
```diff
@@ -177,6 +177,7 @@ type Config struct {
 	EnableAgentTLSForChecks          *bool               `mapstructure:"enable_agent_tls_for_checks" json:"enable_agent_tls_for_checks,omitempty"`
 	EnableCentralServiceConfig       *bool               `mapstructure:"enable_central_service_config" json:"enable_central_service_config,omitempty"`
 	EnableDebug                      *bool               `mapstructure:"enable_debug" json:"enable_debug,omitempty"`
+	EnableDevResources               *bool               `mapstructure:"enable_dev_resources" json:"enable_dev_resources,omitempty"`
```
Comment (mkeeler) [inline agent/config/config.go:180]: I was getting ready to implement a generic feature flag set in the config. Something like:

```
features = [
   "catalog-v2-alpha",
]
```

Do you think this dev resource enablement would fit better into that framework?

File: test/integration/consul-container/libs/cluster/container.go [hunk 4]
```diff
@@ -561,6 +602,7 @@ func newContainerRequest(config Config, opts containerOpts, ports ...int) (podRe
 		ExposedPorts: []string{
 			"8500/tcp", // Consul HTTP API
 			"8501/tcp", // Consul HTTPs API
+			"8502/tcp", // Consul gRPC API
```
Comment (mkeeler) [inline test/integration/consul-container/libs/cluster/container.go:605]: Regarding grpc TLS, wouldn't port 8503 need to be exposed here?

File: agent/config/config.go (hunk 3 above)
Comment (boxofrad) [reply agent/config/config.go:180]: Sounds good to me!

File: test/integration/consul-container/libs/cluster/container.go (hunk 4 above)
Comment (boxofrad) [reply test/integration/consul-container/libs/cluster/container.go:605]: Yep! Didn't include right now, as it wasn't clear how to get the test harness to use gRPC TLS anway.

File: test/integration/consul-container/libs/cluster/container.go (hunk 1 above)
Comment (huikang) [reply test/integration/consul-container/libs/cluster/container.go:242]: @boxofrad , thanks for tagging us. Sure. Will take a look. (cc @NiniOak , @nfi-hashicorp )

File: test/integration/consul-container/test/resources/resources_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+			Kind:         "album",
+		},
+		Tenancy: &pbresource.Tenancy{
+			Partition: "default",
+			PeerName:  "local",
+			Namespace: "default",
+		},
+	})
+	require.NoError(t, err)
+	require.Len(t, listRsp.Resources, 3)
```
Comment (analogue) [inline test/integration/consul-container/test/resources/resources_test.go:85]: Would be nice finish this out with deleting the artist and asserting all the albums have been deleted via cascading deletes.

Comment (boxofrad) [reply test/integration/consul-container/test/resources/resources_test.go:85]: Yep, definitely! 🎉

File: test/integration/consul-container/test/resources/resources_test.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	listRsp, err := client.List(ctx, &pbresource.ListRequest{
+		Type: &pbresource.Type{
+			Group:        "demo",
+			GroupVersion: "v2",
+			Kind:         "album",
+		},
+		Tenancy: writeRsp.Resource.Id.Tenancy,
+	})
+	require.NoError(t, err)
+	require.Len(t, listRsp.Resources, 3)
```
Comment (huikang) [inline test/integration/consul-container/test/resources/resources_test.go:81]: Why there are 3 resources? I think the test case only creates one resource. (Again I am not that familiar with the new model, please point me to any doc that will help me to understand the model. Thanks.)

**Review by huikang — COMMENTED** [review]
Overall looks good; left some questions.

Do you mind rebasing this PR against the latest main as I made some changes to the consul-container to make integ-test work without testing splitting? Not sure how this might impact the current status. Thanks.

File: test/integration/consul-container/test/resources/resources_test.go (hunk 6 above)
Comment (boxofrad) [reply test/integration/consul-container/test/resources/resources_test.go:81]: Here, we're checking that the demo artist controller has created three album resources. In other words, we're testing the controller's side-effect.

https://github.com/hashicorp/consul/blob/113202d541467ef55ca6b6c2204f1b3a669e8648/internal/resource/demo/controller.go#L61-L64

**Comment by boxofrad** [comment]
Heads up @mkeeler and @huikang 😄 ⬆️ 

Now that the node health controller has been merged, I've rewritten this integration test to target that controller instead of the "demo" controller.

Originally, I didn't want to couple this test to a specific Consul feature (e.g. catalog), but I now think it's the most pragmatic option because:
1. Integration tests can't use `-dev` mode, so we'd have to introduce a new flag to enable the demo types/controllers.
2. We have a [rule](https://github.com/hashicorp/consul/blob/113202d541467ef55ca6b6c2204f1b3a669e8648/GNUmakefile#L347-L356) that prevents us from importing the root module in our integration tests. The demo types live in `proto/private` so we'd need to either move them, or perhaps re-generate the protobuf code.

Both of which seem like far too much hassle for such a simple test.

**Pull request closed without merging**